mirgecom/profiling.py

```diff
@@ -0,0 +1,208 @@
+
+            times = [v.time for v in value]
+            flops = [v.flops for v in value]
+            bytes_accessed = [v.bytes_accessed for v in value]
+
+            print(format_str.format(key.name, num_values, min(times),
+                int(mean(times)), max(times), min(flops), int(mean(flops)),
+                max(flops), min(bytes_accessed), int(mean(bytes_accessed)),
+                max(bytes_accessed), round(mean(bytes_accessed)/mean(times), 3)))
+
+        print(format_str.format('='*20, '='*6, '='*8, '='*8, '='*8, '='*8, '='*8,
+            '='*8, '='*8, '='*8, '='*8, '='*8))
+
+    def call_loopy(self, program, **kwargs) -> dict:
+        program = self.transform_loopy_program(program)
+        assert program.options.return_dict
+        assert program.options.no_numpy
+
+        if self.profiling_enabled:
+            args_tuple = ()
+
+            for key, value in kwargs.items():
+                if hasattr(value, 'shape'):
+                    args_tuple = args_tuple + (key, value.shape)
+                else:
+                    args_tuple = args_tuple + (key, value)
+
+            # Check if we need to get the invoker code to generate integer arguments
+            # N.B.: The invoker code might be different for the same program with
+            # different kwargs
+            if (program not in self.kernel_stats
+              or args_tuple not in self.kernel_stats[program]):
+                executor = program.target.get_kernel_executor(program, self.queue)
+                info = executor.kernel_info(executor.arg_to_dtype_set(kwargs))
+
+                kernel = executor.get_typed_and_scheduled_kernel(
+                    executor.arg_to_dtype_set(kwargs))
+
+                data = info.implemented_data_info
+
+                # Generate the wrapper code
+                wrapper = executor.get_wrapper_generator()
+
+                gen = PythonCodeGenerator()
+
+                wrapper.generate_integer_arg_finding_from_shapes(gen, kernel, data)
+                wrapper.generate_integer_arg_finding_from_offsets(gen, kernel, data)
+                wrapper.generate_integer_arg_finding_from_strides(gen, kernel, data)
+
+                types = {}
+                param_dict = {}
+
+                for key, value in kwargs.items():
+                    if hasattr(value, 'dtype') and not value.dtype == object:
+                        types[key] = value.dtype
+                    param_dict[key] = value
+
+                for key, value in kernel.arg_dict.items():
+                    if key not in param_dict:
+                        param_dict[key] = None
+
+                for d in data:
+                    if d.name not in param_dict:
+                        param_dict[d.name] = None
+
+                # Run the wrapper code, save argument values in param_dict
+                exec(gen.get(), param_dict)
+
+                # Get flops/memory statistics
+                kernel.options.ignore_boostable_into = True
```

Comment:
Stateful modification is not allowed, for any data structure in loopy, and for most data structures in code I write.
```
kernel = kernel.copy(
   options=options.copy(ignore_boostable_into=True))
```
is the correct idiom.

But given https://github.com/inducer/loopy/pull/128, this should be a no-op anyway... so I wonder what this is doing here.

Reply:
Note that `kernel` is already a copy of `program`. As you said, this line seems to be not necessary anymore (but was necessary earlier when I wrote the code). Removed.

Reply:
It's still invalid, despite being a copy. `LoopKernel` computes (and stores) a hash key. If you modify it in-place, you'll screw up the haashing logic royally.